# Sprint 3 - Manipulación de Datos I (Sesiones)

En el presente caso de estudio se propone una introducción práctica a los fundamentos de manipulación y análisis de datos utilizando la librería **Pandas**, una de las herramientas más importantes y ampliamente utilizadas dentro del ecosistema de **Python** para el procesamiento de información estructurada. Ante esto, el objetivo principal de estas actividades es desarrollar habilidades iniciales para trabajar con conjuntos de datos reales, comprendiendo no solo las técnicas básicas de transformación y exploración de información, sino también la lógica metodológica que guía un proyecto de análisis de datos.

En el ámbito profesional, el análisis de datos no se limita únicamente a ejecutar código o generar gráficos; requiere comprender el problema que se desea resolver, interpretar correctamente la información disponible y estructurar adecuadamente cada etapa del proceso analítico. Por esta razón, además de introducir herramientas técnicas de manipulación de datos, este caso de estudio busca familiarizarte con algunos de los principios fundamentales que permiten organizar y desarrollar proyectos analíticos de manera ordenada, reproducible y orientada a la toma de decisiones.

Por un lado, a través de **Pandas** trabajarás con estructuras tabulares de información, aprendiendo a cargar, inspeccionar, limpiar, transformar y analizar datos provenientes de archivos externos. Estas capacidades técnicas constituyen la base de disciplinas como la analítica de negocios, la ciencia de datos, la inteligencia artificial y la investigación aplicada, donde la calidad del análisis depende directamente de la correcta preparación y comprensión de los datos disponibles.

De manera complementaria, durante el desarrollo de este caso se introducirán los primeros criterios metodológicos asociados a un flujo de trabajo analítico básico, abordando por ahora las siguientes etapas:

* Entendimiento del contexto.
* Entendimiento de los datos.
* Preparación de los datos.
* Análisis de datos.

Cada una de estas etapas cumple un rol fundamental dentro de un proyecto de análisis de datos. Comprender el contexto permite identificar el problema de negocio que se desea resolver; entender los datos facilita reconocer la utilidad, limitaciones, estructura y significado de los mismos; preparar los datos asegura la consistencia y calidad de la información a usarse; y finalmente, el análisis de datos permite extraer hallazgos, patrones y conclusiones útiles para apoyar procesos de decisión en el ámbito del problema definido.

Importante mencionar que el desarrollo de estas habilidades representa un primer acercamiento a la construcción de proyectos analíticos reales, promoviendo no solo competencias técnicas relacionadas con programación y manipulación de datos, sino también capacidades analíticas y metodológicas esenciales para enfrentar problemas complejos basados en información.

## Entendimiento del contexto

El ranking de selecciones masculinas de la FIFA constituye uno de los sistemas de clasificación deportiva profesional más reconocidos a nivel mundial. Este mecanismo fue implementado oficialmente en el año 1993, poco antes de la realización de la Copa Mundial de 1994, con el objetivo de establecer una metodología estandarizada que permitiera comparar el desempeño relativo de las distintas selecciones nacionales afiliadas. Desde entonces, el ranking se ha convertido en una referencia central para evaluar el rendimiento competitivo de los equipos nacionales en el contexto internacional.

Esta clasificación se actualiza de manera periódica —habitualmente con frecuencia mensual— tomando en consideración los resultados obtenidos por cada selección en partidos oficiales y amistosos reconocidos por la FIFA. Entre los encuentros considerados se incluyen competiciones continentales, eliminatorias mundialistas, torneos internacionales y partidos amistosos disputados durante las fechas oficiales establecidas en el calendario oficial. Dependiendo de la relevancia del torneo, la fortaleza del rival y el resultado obtenido, cada encuentro aporta una determinada cantidad de puntos al cálculo global de cada selección.

Más allá de su carácter informativo para aficionados, periodistas y analistas deportivos, el ranking FIFA posee una importancia estratégica significativa dentro de la organización del fútbol internacional. Actualmente, este sistema constituye uno de los principales criterios utilizados para definir los bombos y grupos en competiciones oficiales de gran relevancia, especialmente en Copa Mundial. Esto implica que una mejor posición en el ranking puede influir directamente en el nivel de dificultad que enfrentará una selección durante la fase inicial de un torneo mundialista, afectando potencialmente sus probabilidades de clasificación a rondas posteriores.

A lo largo de su historia, el ranking ha experimentado múltiples ajustes metodológicos orientados a mejorar la precisión y representatividad del sistema de puntuación. Entre las modificaciones más relevantes se encuentran las implementadas en enero de 1999, agosto de 2006, agosto de 2011 y agosto de 2018. Cada uno de estos cambios respondió a críticas y limitaciones detectadas en las versiones anteriores del modelo, particularmente relacionadas con la sobrevaloración de ciertos partidos amistosos, diferencias regionales en la competitividad y fluctuaciones excesivas en las posiciones del ranking.

La modificación más reciente, implementada en 2018, introdujo un sistema basado en modelos tipo Elo, que se utilizan en otros deportes y disciplinas competitivas. Este enfoque busca reflejar de manera más realista el desempeño de las selecciones mediante ajustes dinámicos en función de los resultados esperados y obtenidos en cada partido. Como consecuencia, el ranking actual presenta una mayor estabilidad y sensibilidad frente a la calidad de los oponentes y la importancia competitiva de cada encuentro.

Desde una perspectiva analítica, el ranking FIFA representa un caso de estudio particularmente interesante para el análisis de datos y la modelación estadística. Su evolución histórica permite explorar patrones de desempeño deportivo, cambios en la competitividad entre regiones, impacto de eventos internacionales y comportamiento temporal de las selecciones nacionales. Además, el conjunto de datos asociado al ranking ofrece oportunidades para aplicar técnicas de limpieza, transformación, análisis exploratorio y visualización de datos utilizando herramientas como **Pandas**.

## Entendimiento de los datos

Lo primero que te sugiero hacer al iniciar cualquier proyecto de análisis de datos es importar las librerías y módulos con los que vas a trabajar. En este caso en concreto vamos cargar: 

* **Pandas** cuyo objetivo es proporcionar funciones y métodos para un eficiente manejo de estructuras tabulares (también conocidos como *dataframes* o *datasets*).
* **Numpy** cuyo objetivo es proporcionar funciones y métodos para análisis numérico.

In [1]:
import pandas as pd # pd es el pseudónimo con el que llamaremos a pandas en todo el documento
import numpy as np # np es el pseudónimo con el que llamaremos a numpy en todo el documento

A continuación, conviene cargar los datos en la variable `df_fifa` mediante la función `read_csv` de **Pandas**. En este caso usaremos la información proveniente del archivo [**fifa_rank.csv**](https://raw.githubusercontent.com/jsaraujo5081/material_curso_data_science/refs/heads/main/datasets/fifa_rank.csv). 

In [2]:
df_fifa = pd.read_csv(
    "https://raw.githubusercontent.com/jsaraujo5081/material_curso_data_science/refs/heads/main/datasets/fifa_rank.csv", 
    sep = "," # El argumento sep no es obligatorio pero sirve para especificar el caracter con que los datos están separados en el archivo
) 

Una buena práctica al momento de cargar los datos es observar sus principales características a fin de establecer un plan de tratamiento acorde al dataset y su contexto. Esto resulta de vital importancia puesto que cada conjunto de datos es único y tiene sus propias particularidades. En concreto, interesa ver al menos lo siguiente:

* Estructura de los datos (número de filas y columnas)
* Nombres de las columnas
* Existencia de valores perdidos en cada columna
* Tipos de variables en cada columna
* Aspectos particulares de cada columna

Con este propósito, utiliza el método `info` para un primer vistazo al dataset.

In [3]:
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Country             57797 non-null  str    
 1   Confederation       54815 non-null  str    
 2   Rank_Date           57797 non-null  str    
 3   Points_Old_Version  40382 non-null  float64
 4   Points_New_Version  57797 non-null  str    
 5   rank                57797 non-null  int64  
dtypes: float64(1), int64(1), str(4)
memory usage: 4.1 MB


Una forma más expresa de conocer las dimensiones del dataset (filas y columnas) es mediante el atributo `shape`. Aplicalo en nuestros datos.

In [4]:
df_fifa.shape

(57797, 6)

El dataset **fifa_rank** contiene cerca de 58 mil filas y 6 columnas detalladas a continuación:

* Country: Nombre del pais afiliado a la FIFA cuya selección de fútbol es puntuada.
* Confederation: Confederación a la que pertenece de acuerdo a la división de la FIFA.
* Rank_Date: Fecha de publicación del ranking FIFA entre marzo de 1993 y junio de 2018.
* Points_Old_Version: Puntos obtenidos conforme metodologías de cálculo anteriores al 2011 
* Ponts_New_Version: Puntos obtenidos conforme la metodología posterior a 2011. 
* rank: Posición en el ranking oficial de la FIFA dados los puntos calculados.

A continuación algunos aspectos detectados:

* Los nombres de las columnas tienen mayúsculas y esto puede ser problemático en las siguientes etapas del proyecto puesto que complicarían de forma innecesaria la consulta de la información.
* Existen valores perdidos en las columnas Confederation y Points_Old_Version.
* El tipo de variable de Points_New_Version es *string* lo cual no hace sentido considerando que deberías ver números.
* El tipo de variable de Rank_Date igualmente es *string*, lo cual tampoco hace sentido puesto que aquí deberías ver fechas. Al respecto de esto, que una fecha se encuentre en su tipología adecuada es coherente ya que usualmente va a interesar extraer sus componentes (i.e año, mes, dias, etc.)

Profundiza en tu conocimiento del dataset utilizando ahora el método `sample` con el argumento `n = 20`, a fin de extraer una muestra aleatoria de los datos de tamaño 15.

In [5]:
df_fifa.sample(n = 20)

,Country,Confederation,Rank_Date,Points_Old_Version,Points_New_Version,rank
50190,Turkmenistan,AFC,2015-06-04,NaN,99.16,173
55622,Yemen,AFC,2017-08-10,NaN,191.51,146
48649,Latvia,UEFA,2014-11-27,NaN,338.78,97
36325,Oman,AFC,2009-11-20,417.0,ND,83
26363,Jamaica,CONCACAF,2005-10-17,623.0,ND,40
26618,FYR Macedonia,UEFA,2005-11-21,490.0,ND,90
37157,Moldova,UEFA,2010-03-31,328.0,ND,87
864,Morocco,CAF,1994-02-15,44.0,ND,28
16852,Madagascar,CAF,2001-10-17,361.0,ND,122
48334,Papua New Guinea,NaN,2014-09-18,NaN,13.18,199


Considera los siguientes aspectos observados:

* En el caso de los valores perdidos de Points_Old_Version, destaca que los mismos parecen estar asociados a fechas en las que se utilizaban metodologías antiguas de la clasificación, lo cual hace sentido en lo referente al nombre de la columna.
* Points_New_Version presenta casos con el texto "ND", lo cual estaría ocasionando que el tipo de variable no sea el correcto. En todo caso, sucede algo similar con lo visto en el punto anterior, estos "valores ausentes" se corresponderían a fechas donde se aplicaban metodologías anteriores.

Mira más a fondo otras columnas. Por ejemplo, observa los valores posibles de Confederation usando los métodos `unique` y luego `value_counts` con el argumento `dropna = False`.

In [6]:
# Observar valores de Confederation con unique
df_fifa["Confederation"].unique()

<ArrowStringArray>
['UEFA', 'CONMEBOL', 'CONCACAF', 'CAF', 'AFC', nan]
Length: 6, dtype: str

In [7]:
# Observar valores de Confederation con value_counts
df_fifa["Confederation"].value_counts(dropna = False)

Confederation
UEFA        14934
CAF         14876
AFC         12481
CONCACAF     9664
NaN          2982
CONMEBOL     2860
Name: count, dtype: int64

Verifica aquellas selecciones nacionales que tienen valores perdidos en Confederation para comprender a qué confederación corresponderían. Usa para esto el método `isna`.

In [8]:
df_fifa[df_fifa["Confederation"].isna()]["Country"].unique()

<ArrowStringArray>
[       'Australia',      'New Zealand',             'Fiji',
           'Tahiti',  'Solomon Islands',          'Vanuatu',
 'Papua New Guinea',            'Tonga',            'Samoa',
     'Cook Islands',   'American Samoa',    'New Caledonia']
Length: 12, dtype: str

De esta observación en particular se confirma que una confederación está faltando que corresponde a la de Oceanía (OFC).

Utiliza ahora el método `describe` para generar un resumen estadístico de la columna rank. 

In [9]:
df_fifa["rank"].describe()

count    57797.000000
mean       101.629514
std         58.618560
min          1.000000
25%         51.000000
50%        101.000000
75%        152.000000
max        209.000000
Name: rank, dtype: float64

Notemos que en algunas mediciones se han alcanzado hasta 209 paises puesto que este es su valor máximo. De hecho, vale mencionar que la FIFA actualmente reconoce a 211 países por lo que podríamos concluir que algunos países comparten una misma posición.

Mantengamos esa curiosidad por conocer más de los datos. Cuenta ahora cuántos paises fueron rankeados en cada una de las fechas. Ordena este conteo por fecha con el método `sort_index` y muestra únicamente los primeros y últimos resultados con los métodos `head` y `tail` respectivamente.

In [10]:
df_fifa["Rank_Date"].value_counts().sort_index().head()

Rank_Date
1993-08-08    167
1993-09-23    167
1993-10-22    167
1993-11-19    168
1993-12-23    168
Name: count, dtype: int64

In [11]:
df_fifa["Rank_Date"].value_counts().sort_index().tail()

Rank_Date
2018-02-15    211
2018-03-15    211
2018-04-12    211
2018-05-17    211
2018-06-07    211
Name: count, dtype: int64

Comprueba finalmente si existen casos duplicados en el dataset. Considerando que existe una gran cantidad de registros en el dataset (aproximadamente 58 mil filas), sería muy poco práctico hacer una comprobación manual, por tanto utiliza el método `duplicated`.

In [12]:
df_fifa.duplicated().sum()

np.int64(41)

Mira en mayor detalle estos duplicados para saber en que consisten.

In [13]:
df_fifa[df_fifa.duplicated()]["Country"].value_counts()

Country
Sudan                   37
Finland                  1
United Arab Emirates     1
Samoa                    1
Brunei Darussalam        1
Name: count, dtype: int64

Gran parte de estos duplicados corresponden a calificaciones asociadas a la selección de Sudán. En general, los valores duplicados pueden considerarse errores en el registro de información y nos interesa eliminar estos casos a fin de evitar distorsiones en los futuros resultados del análisis. 

A partir de todo este entendimiento de los datos, establece un plan de acción para procesar esta información de modo que la tabla final resultante esté lista para cualquier análisis que desees hacer.

**PLAN DE ACCIÓN PARA PREPARACIÓN DE DATOS**

| # | Acción |
| ---- | ---- |
| 1 | Los nombres de las columnas no están bien presentados, en concreto rank esta en minúsculas y el resto no. Se debería estandarizar a un mismo formato que siga el esquema *snake_case* (minúsculas y con espacios representados por "_"). |
| 2 | Las columna Confederation presenta valores ausentes que pueden corregirse inputando el valor "OFC" debido a que se identificó que estos son los países a los que corresponden dichos errores en la información. |
| 3 | La columna Poins_New_Version debe cambiarse a un tipo flotante (*float*) que responde más a su naturaleza de valores numéricos. Con este propósito en primera instancia conviene dar tratamiento a los casos "ND" detectados cambiándolos por valores ausentes de tipo numérico. |
| 4 | Las columnas Points_Old_Version y Point_New_Version igualmente presentan valores ausentes, sin embargo por el contexto del caso éstos responden a los cambios metodológicos generados por la FIFA. En todo caso convendría crear una nueva columna que consolide esta información y que por tanto ya no tenga casos ausentes. |
| 5 | La columna Rank_Date debe cambiarse al tipo fecha (*datetime*). A partir de esto, se podría contar con campos adicionales de años y meses considerando que la calificación de la FIFA mentiene generalmente esta periodicidad. |
| 6 | Existen 41 filas duplicadas que deben eliminarse considerando que son errores en el dataset y afectarían los resultados de futuros análisis. |

## Preparación de los datos

Utilizando nuestro plan de acción, procedamos a dejar listo este dataset para los análisis futuros que vamos a desarrollar. Entonces, en primer lugar ajusta los nombres de las columnas para que estén en formato *snake_case*. Recuerda que este formato exige que todos los nombres se encuentren en minúsculas y los espacios se sustituyan por "_".

In [14]:
nombres_sc = []
for col in df_fifa.columns:
    nombres_sc.append(col.lower().replace(" ","_"))

df_fifa.columns = nombres_sc
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             57797 non-null  str    
 1   confederation       54815 non-null  str    
 2   rank_date           57797 non-null  str    
 3   points_old_version  40382 non-null  float64
 4   points_new_version  57797 non-null  str    
 5   rank                57797 non-null  int64  
dtypes: float64(1), int64(1), str(4)
memory usage: 4.1 MB


Otra forma más elegante y eficiente para hacer esto es mediante una sintaxis llamada *Python Comprehension* que dejaría todo el código anterior así:

```py
df_fifa.columns = [col.lower().replace(" ","_") for col in df_fifa.columns]
```

Como siguiente paso, inputa el valor "OFC" a los casos ausentes de la columna confederation. Utiliza para esto el método `fillna`.

In [15]:
df_fifa["confederation"] = df_fifa["confederation"].fillna("OFC")
df_fifa["confederation"].value_counts(dropna = False)

confederation
UEFA        14934
CAF         14876
AFC         12481
CONCACAF     9664
OFC          2982
CONMEBOL     2860
Name: count, dtype: int64

Ajusta ahora la columna points_new_version realizando los siguientes pasos:

1. Cambia los valores perdidos "ND" por un valor perdido adecuado usando para esto el valor `np.nan`.
2. Cambia el tipo de variable a flotante utilizando el método `astype`.

In [16]:
# Cambiar valores perdidos ND por Nan
df_fifa.loc[df_fifa["points_new_version"] == "ND","points_new_version"] = np.nan
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             57797 non-null  str    
 1   confederation       57797 non-null  str    
 2   rank_date           57797 non-null  str    
 3   points_old_version  40382 non-null  float64
 4   points_new_version  17415 non-null  str    
 5   rank                57797 non-null  int64  
dtypes: float64(1), int64(1), str(4)
memory usage: 4.0 MB


In [17]:
# Cambiar tipo de variable de points_new_version
df_fifa['points_new_version'] = df_fifa['points_new_version'].astype(float)
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             57797 non-null  str    
 1   confederation       57797 non-null  str    
 2   rank_date           57797 non-null  str    
 3   points_old_version  40382 non-null  float64
 4   points_new_version  17415 non-null  float64
 5   rank                57797 non-null  int64  
dtypes: float64(2), int64(1), str(3)
memory usage: 3.9 MB


Consolida las columnas points_old_version y points_new_version en una nueva columna llamada *points*, de forma que ya no existan valores ausentes por los cambios metodológicos.

In [18]:
df_fifa.loc[df_fifa["points_old_version"].isna(), "points"] = df_fifa["points_new_version"]
df_fifa.loc[df_fifa["points_new_version"].isna(), "points"] = df_fifa["points_old_version"]
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             57797 non-null  str    
 1   confederation       57797 non-null  str    
 2   rank_date           57797 non-null  str    
 3   points_old_version  40382 non-null  float64
 4   points_new_version  17415 non-null  float64
 5   rank                57797 non-null  int64  
 6   points              57797 non-null  float64
dtypes: float64(3), int64(1), str(3)
memory usage: 4.4 MB


Para evitar confusiones posteriores, elimina las columnas points_old_version y points_new_version. Utiliza el método `drop`.

In [19]:
df_fifa = df_fifa.drop(columns = ["points_old_version","points_new_version"])
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   country        57797 non-null  str    
 1   confederation  57797 non-null  str    
 2   rank_date      57797 non-null  str    
 3   rank           57797 non-null  int64  
 4   points         57797 non-null  float64
dtypes: float64(1), int64(1), str(3)
memory usage: 3.5 MB


Transforma la columna rank_date a tipo fecha (*datetime*) y aplicando la función `pd.to_datetime`.

In [20]:
df_fifa["rank_date"] = pd.to_datetime(df_fifa["rank_date"])
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   country        57797 non-null  str           
 1   confederation  57797 non-null  str           
 2   rank_date      57797 non-null  datetime64[us]
 3   rank           57797 non-null  int64         
 4   points         57797 non-null  float64       
dtypes: datetime64[us](1), float64(1), int64(1), str(2)
memory usage: 2.9 MB


Ahora que esta columna ya es una fecha, es posible extraer sus componentes fundamentales. En este setido, crea una columna *year* y otra *month* que contenga la información correspondiente. 

In [21]:
# Crear columna year
df_fifa['year'] = df_fifa['rank_date'].dt.year

# Crear columna month
df_fifa['month'] = df_fifa['rank_date'].dt.month
df_fifa.info()

<class 'pandas.DataFrame'>
RangeIndex: 57797 entries, 0 to 57796
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   country        57797 non-null  str           
 1   confederation  57797 non-null  str           
 2   rank_date      57797 non-null  datetime64[us]
 3   rank           57797 non-null  int64         
 4   points         57797 non-null  float64       
 5   year           57797 non-null  int32         
 6   month          57797 non-null  int32         
dtypes: datetime64[us](1), float64(1), int32(2), int64(1), str(2)
memory usage: 3.4 MB


Adicionalmente, y si bien no está en el plan de acción, podría ser útil especificar en una columna nueva la metodología utilizada para el cálculo de puntos. Crea por tanto la columna *method* la cual cumpla con los siguientes criterios secuenciales:

* Si el año y mes de la fila es menor a Enero 1999 entonces su valor será "1993 - 1998".
* Si el año y mes de la fila es menor a Agosto 2006 entonces su valor será "1999 - 2006"
* Si el año y mes de la fila es menor a Agosto 2011 entonces su valor será "2006 - 2011"
* Todos los demás casos deberían tener el valor "2011 - 2018"

In [22]:
# Crear columna method
df_fifa["method"] = ""

# Generar un bucle para recorrer todas las filas del dataset
for i in range(df_fifa.shape[0]):

    periodomes = df_fifa.loc[i,"year"]*100 + df_fifa.loc[i,"month"]
    
    if periodomes < 199901:
        df_fifa.loc[i,"method"] = "1993-1998"
    elif periodomes < 200608:
        df_fifa.loc[i,"method"] = "1999-2006"
    elif periodomes < 201108:
        df_fifa.loc[i,"method"] = "2006-2011"
    else:
        df_fifa.loc[i,"method"] = "2011-2018"

df_fifa["method"].value_counts().sort_index()

method
1993-1998    10058
1999-2006    18113
2006-2011    12211
2011-2018    17415
Name: count, dtype: int64

Seguramente notaste que la ejecución del programa anterior tardó unos segundos. Lo anterior se explica porque el uso de bucles `for` para estos casos es un tanto ineficiente. Una mejor manera para realizar esto es mediante el método `apply`. Te sugiero que investigues un poco sobre esta herramienta de **Pandas**.

Finalmente, da tratamiento a los duplicados eliminándolos con el método `drop_duplicates`.

In [23]:
df_fifa = df_fifa.drop_duplicates().reset_index(drop = True) # En general, siempre es recomendable resetear los índices cuando se borran filas 
df_fifa.duplicated().sum()

np.int64(0)

Listo, hemos concluido con nuestro plan de acción. Dale una mirada nuevamente al dataframe limpio extrayendo una muestra de 15 elementos.

In [24]:
df_fifa.sample(15)

,country,confederation,rank_date,rank,points,year,month,method
31715,Ukraine,UEFA,2008-01-16,30,824.00,2008,1,2006-2011
54237,Haiti,CONCACAF,2017-02-09,69,517.72,2017,2,2011-2018
37620,Turkmenistan,AFC,2010-05-26,139,153.00,2010,5,2006-2011
24083,Greece,UEFA,2004-11-10,17,703.00,2004,11,1999-2006
24768,Albania,UEFA,2005-02-16,87,515.00,2005,2,1999-2006
32254,Suriname,CONCACAF,2008-03-12,155,122.00,2008,3,2006-2011
37605,Grenada,CONCACAF,2010-05-26,124,198.00,2010,5,2006-2011
12185,Malaysia,AFC,1999-11-17,116,317.00,1999,11,1999-2006
15499,Pakistan,AFC,2001-03-14,191,72.00,2001,3,1999-2006
22901,Scotland,UEFA,2004-05-12,63,540.00,2004,5,1999-2006


## Análisis de datos

Es hora que realices tu primer análisis de datos ahora que tienes una tabla con información limpia. Para un correcto análisis vale que tengas siempre en cuenta lo siguiente:

* Su propósito fundamental es **responder preguntas de negocio**, por lo que siempre debes buscar dar soluciones concretas en base a tus resultados.
* No se limita a realizar gráficos o tablas resumen, sino a generar **conclusiones** que aporten valor a partir de estas visualizaciones.
* Debe mantener un **hilo de ideas** conductor lo cual implica orden y consistencia entre los resultados.

Visto esto, a continuación verás algunas preguntas de negocio a responder: 

### ¿Cómo ha sido el comportamiento anual del puntaje de la CONMEBOL a partir de la última metodología?

En primer lugar identifica entre las columnas de la tabla cuál es la métrica relevante que debes obtener:

`A partir de la columna points, la métrica sería el puntaje promedio por año.`

Ahora, especifica si debes realizar filtros en la tabla a partir de algunas condiciones:

`Se debe filtrar la columna confederation manteniendo solamente las filas de "CONMEBOL". Además se debe filtrar la culumna method manteniendo solamente las filas de "2011-2018".`

Para terminar, indica si debes agregar o resumir los datos por alguna de las columnas: 

`Los datos deben agregarse por la columna year.`

Construye el código necesario para responder la pregunta utilizando como criterios los aspectos contestados antes.

In [25]:
puntos_ec = (
    df_fifa
    # filtrar datos
    .query("confederation == 'CONMEBOL' and method == '2011-2018'")
    # Agrupar datos
    .groupby("year")
    # Metrica a utilizar
    ["points"].mean()
)
puntos_ec.round(2) # El método round permite visualizar el resultado sin excesivos decimales

year
2011     818.37
2012     848.17
2013     891.15
2014     920.10
2015     912.78
2016     996.97
2017    1035.54
2018     959.85
Name: points, dtype: float64

Genera dos conclusiones relevantes. Recuerda que una conclusión no se limita a una descripción de lo que obtuviste, sino a implicancias, hipótesis o recomendaciones a partir de estos resultados que aporten a la toma de decisiones.

**CONCLUSIONES**

| # | Conclusión |
| ---- | ---- |
| 1 | Se observa un crecimiento sostenido en el puntaje de las selecciones de CONMEBOL en este periodo lo cual está asociado al mejoramiento de selecciones no tan protagonistas como son Chile y Ecuador. En concreto, esto se confirma con las Copas América ganadas por Chile en los años 2015 y 2016, así como con los éxitos de clubes ecuatorianos a nivel internacional como IDV y LDU, los cuales son la base actual del seleccionado nacional. |
| 2 | A pesar de esto, en el último año se evidencia una reducción significativa en el puntaje medio de CONMEBOL, explicado seguramente por los malos resultados de varias selecciones de esta confederación previo al mundial de Rusia de 2018: Argentina perdió dos partidos importantes en su etapa preparatoria (2 - 4 versus Nigeria y 1 - 6 versus España); Colombia perdió un partido preparatorio con Corea del Sur por (1 - 2) y empató otro contra Australia (0 - 0); Chile y Ecuador quedaron eliminadas en la fase clasificatoria a pesar de estar bien posicionadas y en constante crecimiento los últimos años. Bolivia recibió sanciones por parte de la FIFA producto de asuntos administrativos en su ente federativo nacional. |

### ¿Cual ha sido la confederación peor evaluada en los últimos 3 años registrados?

Establece cuál es la métrica de análisis, los filtros y agrupamientos que necesitas y da una respuesta a la pregunta con tu código.

In [26]:
puntos_x_conf = (
    df_fifa
    .query("year >= 2016")
    .pivot_table(
        columns = "year",
        index = "confederation",
        values = "points",
        aggfunc = "mean"
    )

)
puntos_x_conf.round(2)

year,2016,2017,2018
confederation,,,
AFC,252.83,259.39,258.71
CAF,374.93,367.80,341.17
CONCACAF,290.68,278.14,253.57
CONMEBOL,996.97,1035.54,959.85
OFC,98.86,110.24,105.09
UEFA,644.87,642.80,647.56


Genera una conclusión relevante.

**CONCLUSIONES**

| # | Conclusión |
| ---- | ---- |
| 1 | OFC tiene el peor posicionamiento entre las confederaciones lo cual se explicaría por el bajo nivel de fútbol competitivo y profesional que existe en los paises de Oceanía. Esto se acrecentó incluso más el año 2006 en el que Australia se afilió a la AFC para mantener un mayor nivel en sus enfrentamientos deportivos. |